content/en/blog/_posts/2021-08-11-memory-qos-cgroups-v2/index.md

```diff
@@ -0,0 +1,88 @@
+
+The kubelet may not be able to guarantee memory resources even when you specify these with a `limit` or a `request` for a Pod or for one of its containers. Because the memory could be overcommitted at the host operating system level, memory pages that are backed by non-volatile storage (such as the executable code for your app) could be dropped from RAM due to overall memory pressure, even when your Pod is running with Guaranteed QoS.
+
+Kubernetes 1.22 introduces a new alpha feature that provides quality-of-service for memory resources. This new feature relies on specific cgroup v2 controls: `memory.min` and `memory.high`. The container runtime uses those capabilities to define a guaranteed amount of RAM for the cgroup overall, as well as to throttle memory requests which is overcommitted. Using memory QoS on your nodes can help you to improve application availability. Setting `memory.min` to match the memory request can help to retain the memory requests to the Pods and containers that it applies to, which is important for latency-sensitive workloads. In addition, setting `memory.high` is to control application memory use in case a large number of memory requests causes the node to become unavailable in an instant.
+
+## How does it work?
+It uses two interfaces, `memory.min` and `memory.high` of cgroup v2 memory controller, to work.
+
+<table>
+    <tr>
+        <th style="text-align:center">File</th>
+        <th style="text-align:center">Description</th>
+   </tr>
+   <tr>
+        <td>memory.min</td>
+        <td><code>memory.min</code> specifies a minimum amount of memory the cgroup must always retain, i.e., memory that can never be reclaimed by the system. If the cgroup's memory usage reaches this low limit and can’t be increased, the system OOM killer will be invoked.
+        <br>
+        <br>
+        <i>We map it to <code>requests.memory</code></i>
+        </td>
+   </tr>
+   <tr>
+       <td>memory.high</td>
+       <td><code>memory.high</code> is the memory usage throttle limit. This is the main mechanism to control a cgroup's memory use. If a cgroup's memory use goes over the high boundary specified here, the cgroup’s processes are throttled and put under heavy reclaim pressure. The default is max, meaning there is no limit. 
+       <br>
+       <br>
+       <i>We use a formula to calculate <code>memory.high</code>, depending on <code>limits.memory/node allocatable memory</code> and a memory throttling factor</i>
+       </td>
+   </tr>
+</table>
+
+Both containers and pods are considered. When cgroup v2 unified and MemoryQoS feature gate are both enabled, kubelet sets `requests.memory` to `memory.min` at container and pod level cgroup. And it also sets `memory.high` to throttle container memory overcommit allocation.
```

Comment:
I think this is misleading. The kubelet doesn't set this directly; the kubelet calculates what the limits should be and communicates this to the container runtime via the CRI. See https://github.com/kubernetes/kubernetes/pull/102578